### PR TITLE
add rapidcheck tests for intrusive_list and intrusive_forward_list

### PR DESCRIPTION
**Pull request by lukevalenty**

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -7,7 +7,8 @@ function(add_tests)
             "${name}.cpp"
             LIBRARIES
             warnings
-            stdx)
+            stdx
+            rapidcheck_catch)
```
Comment (elbeno) [inline test/CMakeLists.txt:11]: **issue**: Linking with rapidcheck_catch should not be necessary - the cmake machinery already does this at https://github.com/intel/cicd-repo-infrastructure/blob/main/cmake/test.cmake#L110

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -30,6 +31,7 @@ add_tests(
     function_traits
     intrusive_forward_list
     intrusive_list
+    intrusive_list_properties
```
Comment (elbeno) [inline test/CMakeLists.txt:34]: **suggestion**: I'd rather not separate example-based and property-based tests; can we put them all in one target?

File: test/intrusive_list_properties.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        void run(ListModel const &m, ListSut &sut) const override {
+            rc_assert_equal(m, sut);
+            delete sut.pop_front();
+        }
+
+        void show(std::ostream &os) const override { os << "pop_front()"; }
+    };
+};
+
+TEST_CASE("push_back, pop_front", "[intrusive_list_properties]") {
```
Comment (elbeno) [inline test/intrusive_list_properties.cpp:102]: **question**: Where are the intrusive list nodes kept? Because the list doesn't own them...

Comment (lukevalenty) [reply test/intrusive_list_properties.cpp:102]: they are being dynamically allocated and freed...this is related to why the address sanitizer is failing. 😭 i will need a simple scheme to free all the nodes when a list goes out of scope.

File: test/CMakeLists.txt (hunk 2 above)
Comment (lukevalenty) [reply test/CMakeLists.txt:34]: i can do that

File: test/CMakeLists.txt (hunk 1 above)
Comment (lukevalenty) [reply test/CMakeLists.txt:11]: ok 👍🏽

File: test/intrusive_list_properties.cpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+template <typename T> struct list_element_deallocator {
+    T list{};
+
+    ~list_element_deallocator() {
+        while (!list.empty()) {
+            delete list.pop_front();
+        }
+    }
+
+    void clear() {
```
Comment (elbeno) [inline test/intrusive_list_properties.cpp:30]: **suggestion**: Can we DRY out `clear` and the destructor? They do basically the same job?

**Review by elbeno — APPROVED** [review]

File: test/intrusive_list_properties.cpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+};
+} // namespace
+
+using ListModel = std::list<int>;
+
+template <typename T> struct list_element_deallocator {
+    T list{};
+
+    void clear() {
+        std::list<int_node *> nodes_to_delete{};
```
Comment (elbeno) [inline test/intrusive_list_properties.cpp:25]: **question**: Avoiding an include?

**Pull request merged**